packages/devtools_app/lib/src/screens/memory/panes/diff/controller/item_controller.dart

```diff
@@ -61,29 +75,115 @@ class SnapshotListItem extends DiffListItem with AutoDisposeControllerMixin {
 
   String get name => '$_isolateName-$displayNumber';
 
-  ValueListenable<SnapshotListItem?> get diffWith => _diffWith;
-  final _diffWith = ValueNotifier<SnapshotListItem?>(null);
-  void setDiffWith(SnapshotListItem? value) {
+  ValueListenable<SnapshotInstanceItem?> get diffWith => _diffWith;
+  final _diffWith = ValueNotifier<SnapshotInstanceItem?>(null);
+  void setDiffWith(SnapshotInstanceItem? value) {
     _diffWith.value = value;
-    updateSelectedRecord();
+    _handleSelectionChange();
   }
 
-  final ValueListenable<String?> selectedClassName;
+  final ValueListenable<HeapClassName?> _selectedClass;
+
+  final ValueListenable<ClassOnlyHeapPath?> _selectedPath;
+
+  ValueListenable<SingleClassStats?> get selectedSingleClassStats =>
+      _selectedSingleClassStats;
+  final _selectedSingleClassStats = ValueNotifier<SingleClassStats?>(null);
+
+  ValueListenable<DiffClassStats?> get selectedDiffClassStats =>
+      _selectedDiffClassStats;
+  final _selectedDiffClassStats = ValueNotifier<DiffClassStats?>(null);
+
+  ValueListenable<ClassStats?> get selectedClassStats => _selectedClassStats;
+  final _selectedClassStats = ValueNotifier<ClassStats?>(null);
 
-  ValueListenable<HeapClassStatistics?> get selectedClassStats =>
-      _selectedClassStats;
-  final _selectedClassStats = ValueNotifier<HeapClassStatistics?>(null);
+  /// Selected retaining path.
+  ValueListenable<StatsByPathEntry?> get selectedPathEntry =>
+      _selectedPathEntry;
+  final _selectedPathEntry = ValueNotifier<StatsByPathEntry?>(null);
 
   @override
   bool get hasData => heap != null;
 
-  HeapStatistics get statsToShow {
+  HeapClasses classesToShow() {
     final theHeap = heap!;
     final itemToDiffWith = diffWith.value;
-    if (itemToDiffWith == null) return theHeap.stats;
-    return diffStore.compare(theHeap, itemToDiffWith.heap!).stats;
+    if (itemToDiffWith == null) return theHeap.classes;
+    return _diffStore.compare(theHeap, itemToDiffWith.heap!);
   }
 
-  void updateSelectedRecord() => _selectedClassStats.value =
-      statsToShow.statsByClassName[selectedClassName.value];
+  void _handleSelectionChange() {
+    final className = _selectedClass.value;
+    if (className == null) return;
+
+    final heapClasses = classesToShow();
+
+    if (heapClasses is SingleHeapClasses) {
+      _selectedSingleClassStats.value =
+          _selectedClassStats.value = heapClasses.classesByName[className];
```

Comment:
store `heapClasses.classesByName[className]` in local before this if / else statement so we don't have to duplicate

Reply:
we need type promotion here to avoid casting